### PR TITLE
OTP-20 compatibility workaround for queue directory names. 3.6 only!

### PR DESCRIPTION
**Pull request by hairyhum**
Workaround for #1243 for 3.6
OTP-20 has different binary format for atoms (since OTP-14337)
So term_to_binary generates different value.
testm_to_binary_compat now have a function to generate a binary
identical to pre-20 versions.

Does not address management or stomp issues.

### CODE REVIEW

File: src/term_to_binary_compat.erl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+%% Copyright (c) 2017 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(term_to_binary_compat).
+
+-include("rabbit.hrl").
+
+-export([queue_name_to_binary/1]).
+
+queue_name_to_binary(#resource{kind = queue} = {resource, VHost, queue, Name}) ->
```
Comment (michaelklishin) [inline src/term_to_binary_compat.erl]: Should we use `#resource` here to make things a bit more forward-compatible?

File: src/term_to_binary_compat.erl [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    VHostBSize = byte_size(VHost),
+    NameBSize = byte_size(Name),
+    <<131,
```
Comment (michaelklishin) [inline src/term_to_binary_compat.erl]: You floated a version of this with comments next to each byte on Slack. I found them very useful.

File: src/term_to_binary_compat.erl (hunk 1 above)
Comment (hairyhum) [inline src/term_to_binary_compat.erl]: It cannot and should not be forward-compatible.
The function is a workaround and in 3.7 we should get rid of it. The code assumes that the record will be a tuple of this specific format. The match asserts that we haven't changed the record and the function still makes sense.

Comment (michaelklishin) [inline src/term_to_binary_compat.erl]: Makes sense.

**Comment by michaelklishin** [comment]
I'm getting the following test failures (on 19.3):

```
cluster_SUITE > cluster_tests > from_cluster_node1 > refresh_events
    #1. {error,
            {{badarg,
                 [{erlang,byte_size,
                      ["cluster_tests/from_cluster_node1/refresh_events-q"],
                      []},
                  {term_to_binary_compat,queue_name_to_binary,1,
                      [{file,"src/term_to_binary_compat.erl"},{line,25}]},
                  {rabbit_queue_index,queue_name_to_dir_name,1,
                      [{file,"src/rabbit_queue_index.erl"},{line,656}]},
                  {rabbit_queue_index,blank_state,1,
                      [{file,"src/rabbit_queue_index.erl"},{line,539}]},
                  {rabbit_queue_index,init,3,
                      [{file,"src/rabbit_queue_index.erl"},{line,303}]},
                  {rabbit_variable_queue,init,6,
                      [{file,"src/rabbit_variable_queue.erl"},{line,504}]},
                  {rabbit_priority_queue,init,3,
                      [{file,"src/rabbit_priority_queue.erl"},{line,148}]},
                  {rabbit_amqqueue_process,init_it2,3,
                      [{file,"src/rabbit_amqqueue_process.erl"},{line,195}]}]},
             {gen_server2,call,[<16201.924.0>,{init,new},infinity]}}}

cluster_SUITE > cluster_tests > from_cluster_node2 > refresh_events
    #1. {error,
            {{badarg,
                 [{erlang,byte_size,
                      ["cluster_tests/from_cluster_node2/refresh_events-q"],
                      []},
                  {term_to_binary_compat,queue_name_to_binary,1,
                      [{file,"src/term_to_binary_compat.erl"},{line,25}]},
                  {rabbit_queue_index,queue_name_to_dir_name,1,
                      [{file,"src/rabbit_queue_index.erl"},{line,656}]},
                  {rabbit_queue_index,blank_state,1,
                      [{file,"src/rabbit_queue_index.erl"},{line,539}]},
                  {rabbit_queue_index,init,3,
                      [{file,"src/rabbit_queue_index.erl"},{line,303}]},
                  {rabbit_variable_queue,init,6,
                      [{file,"src/rabbit_variable_queue.erl"},{line,504}]},
                  {rabbit_priority_queue,init,3,
                      [{file,"src/rabbit_priority_queue.erl"},{line,148}]},
                  {rabbit_amqqueue_process,init_it2,3,
                      [{file,"src/rabbit_amqqueue_process.erl"},{line,195}]}]},
             {gen_server2,call,[<16202.1365.0>,{init,new},infinity]}}}
```

**Comment by hairyhum** [comment]
Fixed tests, which were using strings instead of binaries for queue names.

**Pull request merged**